actionpack/lib/action_dispatch/middleware/flash.rb

```diff
@@ -174,11 +176,25 @@ def key?(name)
         @flashes.key? name.to_s
       end
 
+      # Deletes the entry for the given key and returns its associated value.
+      #
+      #   flash[:alert] = "Something went wrong."
+      #   flash.delete :alert # => "Something went wrong."
+      #   flash[:alert] # => nil
+      #
       def delete(key)
         key = key.to_s
         @discard.delete key
-        @flashes.delete key
-        self
+        @flashes.delete(key).then do |value|
+          if flash_hash_delete_returns_value
+            value
+          else
+            ActiveSupport::Deprecation::DeprecatedObjectProxy.new(self, <<~WARNING.squish)
+              Rails 7.0 will return the deleted value instead of the FlashHash instance.
+              Set `config.action_dispatch.flash_hash_delete_returns_value = false` to remove the deprecation.
```

Comment:
I think we usually first deprecate behaviour, then we allow people to migrate to this new behaviour, then we remove the old behaviour in the next release.
https://github.com/rails/rails/blob/main/activesupport/lib/active_support/time_with_zone.rb#L45-L50
```suggestion
              Rails 7.1 will return the deleted value instead of the FlashHash instance.
              Set `config.action_dispatch.flash_hash_delete_returns_value = true` to opt into the new behaviour and remove the deprecation.
 ```
```